File: model-value/src/main/java/com/bnorm/barkeep/model/BarValue.java
// Copyright 2016 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.model;

import javax.annotation.Nullable;

import org.jetbrains.annotations.NotNull;

import com.google.auto.value.AutoValue;
import com.google.common.collect.ImmutableSet;
import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.Moshi;

@AutoValue
public abstract class BarValue implements Bar {

  public static Builder builder() {
    return new AutoValue_BarValue.Builder();
  }

  public static JsonAdapter<BarValue> jsonAdapter(Moshi moshi) {
    return new AutoValue_BarValue.MoshiJsonAdapter(moshi);
  }

  @Override
  @Nullable
  public abstract ImmutableSet<Ingredient> getIngredients();

  @Override
  public int compareTo(@NotNull Bar other) {
    return Bar.Companion.getCOMPARATOR().compare(this, other);
  }

  @AutoValue.Builder
  public abstract static class Builder {

    public abstract BarValue build();

    public abstract Builder setId(Long id);

    public abstract Builder setTitle(String title);

    public abstract Builder setDescription(String description);

    public abstract Builder setOwner(User owner);

    abstract ImmutableSet.Builder<Ingredient> ingredientsBuilder();

    public Builder addIngredient(Ingredient ingredient) {
      ingredientsBuilder().add(ingredient);
      return this;
    }
  }
}

File: model-value/src/main/kotlin/com/bnorm/barkeep/model/BarValue.kt
// Copyright 2016 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.model

import com.squareup.moshi.FromJson
import com.squareup.moshi.JsonAdapter
import com.squareup.moshi.ToJson

data class BarSpecValue(override val title: String?,
                        override val description: String? = null,
                        override val owner: User? = null) : BarSpec

data class BarValue(override val id: Long = -1,
                    override val title: String?,
                    override val description: String? = null,
                    override val owner: User? = null,
                    override val ingredients: Set<Ingredient> = emptySet()) : Bar

object BarValueAdapter {
  @ToJson fun toJson(bar: BarSpec): BarSpecValue {
    return BarSpecValue(bar.title, bar.description, bar.owner)
  }

  @FromJson fun fromJson(json: BarSpecValue): BarSpec {
    return json
  }

  @ToJson fun toJson(bar: Bar): BarValue {
    return BarValue(bar.id, bar.title, bar.description, bar.owner, bar.ingredients)
  }

  @FromJson fun fromJson(json: BarValue): Bar {
    return json
  }
}
